layers/+lang/reason/packages.el

```diff
@@ -0,0 +1,160 @@
+      (spacemacs/declare-prefix-for-mode 'reason-mode "mr" "refactor")
+      (spacemacs/declare-prefix-for-mode 'reason-mode "m=" "refmt")
+
+      (spacemacs/set-leader-keys-for-major-mode 'reason-mode
+        "cr" 'refmt
+        "==" 'refmt
+        "tr" 'spacemacs/toggle-reason-auto-refmt
+        "=mr" 'reason/refmt-ml-to-re
+        "=rm" 'reason/refmt-re-to-ml))))
+
+(defun reason/post-init-evil-matchit ()
+  (evilmi-load-plugin-rules '(reason-mode) '(template simple html))
+  (add-hook 'reason-mode-hook 'turn-on-evil-matchit-mode))
+
+(defun reason/pre-init-popwin ()
+  (spacemacs|use-package-add-hook popwin
+    :post-config
+    (push '("*Refmt Errors*" :tail t :position bottom :noselect t) popwin:special-display-config)))
+
+(defun reason/pre-init-utop ()
+  (spacemacs|use-package-add-hook utop
+    :post-init
+    (add-hook
+     'reason-mode-hook
+     (lambda ()
+       (setq utop-command "rtop -emacs")
+       (setq utop-edit-command nil)
+       (setq utop-prompt 'reason/rtop-prompt)
+       (setq utop-initial-command "let myVar = \"Hello Reason!\";")
+       (setq utop-phrase-terminator ";")
+       ))
+    :post-config
+    (progn
+      (spacemacs/set-leader-keys-for-major-mode 'reason-mode
+        "er" 'utop-eval-region
+        "eb" 'utop-eval-buffer
+        "ee" 'utop-eval-phrase)
+      )
+    ))
+
+(defun reason/post-init-merlin ()
+  (use-package merlin
+    :defer t
+    :init
+    (progn
+      (setq merlin-completion-with-doc t)
+
+      (spacemacs/set-leader-keys-for-major-mode 'reason-mode
+        "cp" 'merlin-project-check
+        "cv" 'merlin-goto-project-file
+        "eC" 'merlin-error-check
+        "en" 'merlin-error-next
+        "eN" 'merlin-error-prev
+        "gb" 'merlin-pop-stack
+        "gg" 'merlin-locate
+        "gG" 'spacemacs/merlin-locate-other-window
+        "gl" 'merlin-locate-ident
+        "gi" 'merlin-switch-to-ml
+        "gI" 'merlin-switch-to-mli
+        "go" 'merlin-occurrences
+        "hh" 'merlin-document
+        "ht" 'merlin-type-enclosing
+        "hT" 'merlin-type-expr
+        "rd" 'merlin-destruct)
+      )))
+
+(defun flycheck-ocaml-reason-setup ()
+  (with-eval-after-load 'merlin
+    (setq merlin-error-after-save nil)
+
+    (flycheck-define-generic-checker 'reason-merlin
+      "A syntax checker for Reason using Merlin Mode.
+
+    See URL `https://github.com/the-lambda-church/merlin'."
+      :start #'flycheck-ocaml-merlin-start
+      :verify #'flycheck-verify-ocaml-merlin
+      :modes '(reason-mode)
+      :predicate (lambda () (and merlin-mode
+                                 ;; Don't check if Merlin's own checking is
+                                 ;; enabled, to avoid duplicate overlays
+                                 (not merlin-error-after-save))))
+
+    (interactive)
+    (add-to-list 'flycheck-checkers 'reason-merlin)))
+
+(when (configuration-layer/layer-used-p 'syntax-checking)
```

Comment:
#### First:
Each package that's defined for a layer, probably needs to have an init function defined. Therefore the following functions can't be wrapped in this when statement.

That might be why I get these warning messages on startup:
>Warnings:
    - package flycheck not initialized in layer reason, you may consider removing this
      package from the package list or use the :toggle keyword instead of a \`when'
      form.
    - package flycheck-ocaml not initialized in layer reason, you may consider
      removing this package from the package list or use the :toggle keyword instead
      of a \`when' form.
    - Cannot find "opam" or "merlin" executable. The ocaml layer won't work properly.

The last line is most likely unrelated, I just haven't installed `opam` or `merlin`.

#### Second:
There's something strange going on with this line.
```elisp
(when (configuration-layer/layer-used-p 'syntax-checking)
```

It returns `nil` even though `syntax-checking` is listed in the `dotspacemacs-configuration-layers` variable.

This can be seen by adding these lines above the when statement:
```elisp
(message "dotspacemacs-configuration-layers: %s"
         dotspacemacs-configuration-layers)
(message "(configuration-layer/layer-used-p 'ocaml): %s"
         (configuration-layer/layer-used-p 'ocaml))
(message "(configuration-layer/layer-used-p 'syntax-checking): %s"
         (configuration-layer/layer-used-p 'syntax-checking))
```

When Spacemacs is restarted, then the messages buffer shows:
```
dotspacemacs-configuration-layers: (autohotkey colors emacs-lisp git helm markdown multiple-cursors ocaml org reason spell-checking syntax-checking treemacs version-control)
(configuration-layer/layer-used-p ’ocaml): (ocaml multiple-cursors markdown helm git emacs-lisp colors autohotkey)
(configuration-layer/layer-used-p ’syntax-checking): nil
```

`ocaml` is found but not `syntax-checking`

---



Each init function that requires the `syntax-checking` layer might need to check if it exists individually.

Same thing with `(configuration-layer/layer-used-p 'ocaml)` below.

#### System Info :computer:
- OS: windows-nt
- Emacs: 26.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 642eaf11f), **This PR has been applied on: b80154bd52b83b0877c6de24b0a73b48661f5ec5**
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey colors emacs-lisp git helm markdown multiple-cursors ocaml org reason spell-checking syntax-checking treemacs version-control)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS LCMS2

Reply:
For your second point: I think I've run into something similar before - the order of the layers matters, so syntax-checking has to come before reason in your layer list for this to work.

I'll have a play around with this syntax-checking check anyway, I borrowed it from the ocaml layer but it looks like other layers dont do this the same way, will see if I can simplify.